pkg/controllers/managementuser/rbac/roletemplates/prtb_handler.go

```diff
@@ -0,0 +1,224 @@
+		if hasPromotedRule && matchingPromotedRB == nil {
+			if _, err := p.rbClient.Create(promotedRB); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	// Ensure a service account impersonator exists on the cluster
+	if prtb.UserName != "" {
+		if err := p.impersonationHandler.ensureServiceAccountImpersonator(prtb.UserName); err != nil {
+			return nil, fmt.Errorf("error deleting service account impersonator: %w", err)
+		}
+	}
+
+	return prtb, nil
+}
+
+// OnRemove removes all Role Bindings in each project namespace made by the PRTB
+func (p *prtbHandler) OnRemove(key string, prtb *v3.ProjectRoleTemplateBinding) (*v3.ProjectRoleTemplateBinding, error) {
+	// TODO how to select namespaces
+	namespaces, err := p.nsClient.List(metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	lo := metav1.ListOptions{
+		LabelSelector: createPRTBOwnerLabel(prtb.Name),
+	}
+
+	var returnError error
+	for _, n := range namespaces.Items {
+		rbs, err := p.rbClient.List(n.Name, lo)
+		if err != nil {
+			return nil, err
+		}
+		for _, crb := range rbs.Items {
+			err = p.rbClient.Delete(n.Name, crb.Name, &metav1.DeleteOptions{})
+			if err != nil && !apierrors.IsNotFound(err) {
+				returnError = errors.Join(returnError, err)
+			}
+		}
+	}
+
+	if prtb.UserName != "" {
+		if err = p.impersonationHandler.deleteServiceAccountImpersonator(prtb.UserName); err != nil {
+			return nil, err
+		}
+	}
+
+	return nil, returnError
+}
+
+// doesRoleTemplateHavePromotedRules checks if the PRTB's RoleTemplate has a ClusterRole for promoted rules.
+func (p *prtbHandler) doesRoleTemplateHavePromotedRules(prtb *v3.ProjectRoleTemplateBinding) (bool, error) {
+	rt, err := p.rtClient.Get(prtb.RoleTemplateName, metav1.GetOptions{})
+	if err != nil {
+		return false, err
+	}
+
+	_, err = p.crClient.Get(rbac.PromotedClusterRoleNameFor(rt.Name), metav1.GetOptions{})
+	if err != nil && !apierrors.IsNotFound(err) {
+		return false, err
+	}
+	return !apierrors.IsNotFound(err), nil
+}
+
+// buildRoleBindings returns the Role Binding needed for a prtb in a specific namespace.
+// If createPromotedRoleBindings is true, it also returns a Role Binding for the promoted Cluster Role.
+func buildRoleBindings(prtb *v3.ProjectRoleTemplateBinding, ns string, createPromotedRoleBinding bool) (*v1.RoleBinding, *v1.RoleBinding, error) {
+	ownerLabel := createPRTBOwnerLabel(prtb.Name)
+	roleRef := v1.RoleRef{
+		Kind: "Role",
+		Name: rbac.AggregatedClusterRoleNameFor(prtb.RoleTemplateName),
+	}
+
+	subject, err := rbac.BuildSubjectFromRTB(prtb)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	rb := &v1.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "rb-",
+			Namespace:    ns,
+			Labels:       map[string]string{ownerLabel: "true"},
+		},
+		RoleRef:  roleRef,
+		Subjects: []v1.Subject{subject},
+	}
+
+	var promotedRB *v1.RoleBinding
+	if createPromotedRoleBinding {
+		roleRef.Name = rbac.AggregatedClusterRoleNameFor(rbac.PromotedClusterRoleNameFor(prtb.RoleTemplateName))
```

Comment:
It may seem like you are trying to change the `roleRef` both for `rb` and `promotedRB`, but since it's passed by value and not by pointer, that is not case.
Why not creating a new `roleRef` just here? Or better, just embed it in the `promotedRB` definition.